src/main/java/com/recurse/portfolio/web/UserController.java

```diff
@@ -118,10 +122,29 @@ public RedirectView postEditMyProfile(
             throw new VisibilityException(Visibility.PRIVATE);
         }
 
-        updateMutableCurrentUserValues(currentUser, postedUser);
-        repository.save(currentUser);
+        List<String> errors = validate(postedUser);
+        if (errors.isEmpty()) {
+            updateMutableCurrentUserValues(currentUser, postedUser);
+            repository.save(currentUser);
+
+            return new ModelAndView(new RedirectView("/user/" + userId));
+        } else {
+            postedUser.setUserId(currentUser.getUserId());
+            return new ModelAndView("users/edit")
+                .addObject("user", postedUser)
+                .addObject("errors", errors);
+        }
+    }
 
-        return new RedirectView("/user/" + userId);
+    private List<String> validate(User postedUser) {
+        List<String> errors = new ArrayList<>();
+        if (StringUtils.isEmpty(postedUser.getPublicName())) {
+            errors.add("publicName");
+        }
+        if (StringUtils.isEmpty(postedUser.getInternalName())) {
+            errors.add("internalName");
+        }
+        return errors;
```

Comment:
It turns out [Spring supports form input validation natively](https://spring.io/guides/gs/validating-form-input/), so that we don't have to write this `validate` method in each controller. I'll work on adding that.